feat: instance "name", test ids, store pipes in stdpath(state)

justinmk opened the pull request:
## Problem:

- Unix sockets are created in random /tmp dirs.
  - /tmp is messy, unclear when OSes actually clear it.
  - The generated paths are very ugly. This adds friction to reasoning   about which paths belong to which Nvim instances.
- No way to provide a human-friendly way to identify Nvim instances in logs or server addresses.
- No way to correlate tests with log messages.

## Solution:

- Store unix sockets in stdpath('state'). Closes #3517
- Allow --listen "name" and serverstart("name") to given a name (which is appended to a generated path).
- Generate a test-id for each test and set that as the "name" of Nvim instances spawned from that test.


### Details

Each test gets a test id which looks like "T123". This also appears in the log file. Child processes spawned from a test appear in the logs with the *parent* name followed by "/c". Example:

    DBG 2022-06-15T18:37:45.226 T57.58016.0   UI: flush
    DBG 2022-06-15T18:37:45.226 T57.58016.0   inbuf_poll:442: blocking... events_enabled=0 events_pending=0
    DBG 2022-06-15T18:37:45.227 T57.58016.0/c UI: stop
    INF 2022-06-15T18:37:45.227 T57.58016.0/c os_exit:595: Nvim exit: 0
    DBG 2022-06-15T18:37:45.229 T57.58016.0   read_cb:118: closing Stream (0x7fd5d700ea18): EOF (end of file)
    INF 2022-06-15T18:37:45.229 T57.58016.0   on_process_exit:400: exited: pid=58017 status=0 stoptime=0

## TODO:

- is stdpath(state) the right place to store Nvim sockets? #3517 implies not. I'm hoping someone will raise the alarm if there is a security issue with storing sockets in `~/.local/state/`.

justinmk committed:
feat(logging): include test-id in log messages

Problem:
1. Log messages (especially in CI) are hard to correlate with tests.
2. Since b353a5c05f02 #11886, dumplog() prints the logs next to test
   failures. This is noisy and gets in the way of the test results.

Solution:
1. Associate an incrementing id with each test and include it in log
   messages.
    - FUTURE: add v:name so Nvim instances can be formally "named"?
2. Mention "child" in log messages if the current Nvim is a child (based
   on the presence of $NVIM).

BEFORE:

    DBG … 12345      UI: event
    DBG … 12345      log_server_msg:722: RPC ->ch 1: …
    DBG … 12345      UI: flush
    DBG … 12345      inbuf_poll:444: blocking... events_enabled=1 events_pending=0
    DBG … 23454      UI: stop
    INF … 23454      os_exit:594: Nvim exit: 0

AFTER:

    DBG … T57        UI: event
    DBG … T57        log_server_msg:722: RPC ->ch 1: …
    DBG … T57        UI: flush
    DBG … T57        inbuf_poll:444: blocking... events_enabled=1 events_pending=0
    DBG … T57/child  UI: stop
    INF … T57/child  os_exit:594: Nvim exit: 0

kylo252 commented on src/nvim/log.c:
```diff
@@ -284,13 +287,17 @@ static bool v_do_log_to_file(FILE *log_file, int log_level, const char *context,
+  // TODO(justinmk): introduce v:name?
+  static char name[16] = { 0 };
+
   static const char *log_levels[] = {
```
I think it would be useful to define this list globally, so it can be re-used by the lua logger

https://github.com/neovim/neovim/blob/497046466e4a969cefce5fc2b7992dde157aa1a5/runtime/lua/vim/_editor.lua#L56-L65

and on that note, maybe it would then make sense to keep the name more verbose, e.g. `LOG_LEVEL_x`, to keep the two loggers more in sync.

justinmk replied:
out of scope here, will be dealt with in https://github.com/neovim/neovim/pull/7062

the Lua logger is an accident. what do you mean "globally"? FFI is not an
option, we support puc Lua

kylo252 replied:
> out of scope here, will be dealt with in #7062

thanks for clarifying! 

> what do you mean "globally"? FFI is not an option, we support puc Lua

it seems like the idea is that `log.lua` should be entirely scrapped, in that case this doesn't really matter, but I'm curious, is it not possible to set `vim.log.levels` through `src/nvim/lua/stdlib.c`? sorry if it's a stupid question :)

justinmk replied:
> , is it not possible to set `vim.log.levels` through `src/nvim/lua/stdlib.c`? sorry if it's a stupid question :)

could, but not worth the extra code. once we have a public interface, the log levels will be covered by tests, so the small amount of duplication is less costly than the extra code.

justinmk committed:
test(report): formatting, drop dumplog()

Don't need to dumplog() on each failed test because we now have test-ids
that associate log messages with tests.

justinmk committed:
feat(server): instance "name", store pipes in stdpath(state)

Problem:
- Unix sockets are created in random /tmp dirs.
  - /tmp is messy, unclear when OSes actually clear it.
  - The generated paths are very ugly. This adds friction to reasoning
    about which paths belong to which Nvim instances.
- No way to provide a human-friendly way to identify Nvim instances in
  logs or server addresses.

Solution:
- Store unix sockets in stdpath('state')
- Allow --listen "name" and serverstart("name") to given a name (which
  is appended to a generated path).

TODO:
- is stdpath(state) the right place?

kylo252 commented on test/busted/outputHandlers/nvim.lua:
```diff
@@ -193,6 +194,9 @@ return function(options)
     io.write(globalTeardown)
     io.write(suiteEndString:format(testCount, tests, fileCount, files, elapsedTime_ms))
     io.write(getSummaryString())
+    if failureCount > 0 or errorCount > 0 then
+      io.write(global_helpers.read_nvim_log(nil, true))
+    end
```
isn't it better to have this in either `handler.testEnd()` or `handler.error()`?

otherwise, I don't see much use of dumping the tail of the log. however, it would work if `read_nvim_log()` was grepping for `test_id`

justinmk replied:
1. We've tried testEnd() (that's the current state of `master`). It is too noisy. This PR intentionally avoids that.
2. The tail should be short because most tests don't log (CI sets loglevel=WARN)
3. The logs in theory may be useful because of the test ids

> if `read_nvim_log()` was grepping for `test_id`

That could be interesting. At this point the idea is just to see if it's useful at all, and then improve things if **and only if** need becomes clear.

justinmk commented:
Merging this but will continue iterating. Especially interested in strong arguments against storing sockets / named pipes in `~/.local/state/`.

https://wiki.archlinux.org/title/XDG_Base_Directory says:

> XDG_RUNTIME_DIR
> Used for non-essential, user-specific data files such as sockets, named pipes, etc.

This would be a complication, but if we touch it periodically the socket won't be removed. https://github.com/neovim/neovim/pull/17093#issuecomment-1015286489

> Must be on the local filesystem.

That might be an advantage of XDG_RUNTIME_DIR, if it implies that XDG_STATE_DIR is potentially non-local.

kylo252 commented:
> That might be an advantage of XDG_RUNTIME_DIR, if it implies that XDG_STATE_DIR is potentially non-local.

For the record, `$XDG_RUNTIME_DIR` is managed by `pam_systemd`, so any (Linux) platform that doesn't use systemd and/or [logind ](https://www.freedesktop.org/software/systemd/man/systemd-logind.service.html#) will have a problem with this.

Rahix commented:
> Especially interested in strong arguments against storing sockets / named pipes in `~/.local/state/`

I would not put runtime files like sockets & named pipes into anything beneach `$HOME` for two reasons:

1. Because it is part of the home directory, it is potentially non-local.  Some people have a setup where the home directory is mounted via NFS on multiple hosts and each should probably have separate nvim runtime files.
2. Because it is a persistent location.  There is no value in storing these kind of files on a filesystem backed by a real disk.  A `tmpfs` location is much cleaner here, I would say.

From my understanding `$XDG_STATE_HOME` is meant for _persistent_ state instead of runtime files.  Similar to how a database would store its data in `/var/lib/`, this is the per-user equivalent.  The same database would then maybe have a connection socket in `/run/` - and for this the per-user equivalent is `$XDG_RUNTIME_DIR`.

> For the record, $XDG_RUNTIME_DIR is managed by pam_systemd, so any (Linux) platform that doesn't use systemd and/or logind will have a problem with this.

The standard explicitly suggests to implement a fallback to `/tmp` or an equivalent when `$XDG_RUNTIME_DIR` is not available.  I think this should cover any non-systemd platforms.

justinmk commented:
@Rahix Thank you. In https://github.com/neovim/neovim/pull/18993 I am working on XDG_RUNTIME_DIR support and will use that for storage of sockets, and this will be under /tmp by default.
